### PR TITLE
feat: slashing w/ magnitude metering

### PR DESCRIPTION
**Pull request by 8sunyuan**
PR for magnitude allocation/deallocation and concentration/dilution. Note that magnitude configuration and slashing functions are both in `Slasher.sol`. Could even be collapsed into AVSDirectory depending on contract size but TBD

Implementation TODO
- [X] queueAllocation
- [X] queueDeallocation
- [x] queueMagnitudeConcentration
- [x] queueMagnitudeDilution
- [ ] Incorporating slashing to updates
- [ ] View functions
- [ ] All edge cases TODOs in Slasher

### CODE REVIEW

File: src/contracts/core/Slasher.sol [hunk 1]
```diff
@@ -5,233 +5,475 @@ import "./SlasherStorage.sol";
...
+     * @notice this struct is used in MagnitudeAdjustmentParam in order to specify
+     * an operator's slashability for a certain operator set
+     * @param operatorSet the operator set to change slashing parameters for
+     * @param magnitudeDiff magnitude difference; the difference in proportional parts of the operator's slashable stake
+     * that the operator set is getting. This struct is used either in allocating or deallocating
+     * slashable stake to an operator set. Slashable stake for an operator set is
+     * (slashableMagnitude / sum of all slashableMagnitudes for the strategy/operator + nonSlashableMagnitude) of
+     * an operator's delegated stake.
+     */
+    struct OperatorSetMagnitudeParam {
```
Comment (gpsanant) [inline src/contracts/core/Slasher.sol:29]: can u use the interface pls

File: src/contracts/core/Slasher.sol [hunk 2]
```diff
@@ -5,233 +5,475 @@ import "./SlasherStorage.sol";
...
+    );
+
+    /// operator => strategy => avs => operatorSetId => MagnitudeUpdate[]
+    mapping(address => mapping(IStrategy => mapping(address => mapping(uint32 => MagnitudeUpdate[])))) private _magnitudeUpdates;
+
+    /// operator => strategy => TotalAndNonslashableUpdate[]
+    mapping(address => mapping(IStrategy => TotalAndNonslashableUpdate[])) private _totalMagnitudeUpdates;
+    /// operator => allocator
+    /// could perhaps use 
+    mapping(address => address) public allocatorFor;
```
Comment (gpsanant) [inline src/contracts/core/Slasher.sol:104]: edit delegation manager here according to spec

File: src/contracts/core/Slasher.sol [hunk 3]
```diff
@@ -5,233 +5,475 @@ import "./SlasherStorage.sol";
...
-            "Slasher.increaseRequestedBipsToSlash: bipsToIncrease must be less than BIPS_FACTOR"
-        );
-        _modifyRequestedBipsToSlash(
-            operator, operatorSetID, strategies, EpochUtils.currentEpoch(), int32(bipsToIncrease)
-        );
+        MagnitudeAdjustmentParam[] calldata deallocationParams,
+        SignatureWithExpiry calldata allocatorSignature // TODO: implement signature verification
+    ) external returns(uint32 effectTimestamp) {
+        // perform allocator signature verification if not allocator or operator
+        if (msg.sender != operator && msg.sender != getAllocatorFor(operator)) {
```
Comment (gpsanant) [inline src/contracts/core/Slasher.sol:134]: needs to be allocator

Comment (gpsanant) [reply src/contracts/core/Slasher.sol:134]: operator can only call it if they are also allcoator

File: src/contracts/core/Slasher.sol [hunk 4]
```diff
@@ -5,233 +5,475 @@ import "./SlasherStorage.sol";
...
-     * @param operatorSetID the id of the operator set the AVS is increasing their slashing for
-     * @param strategies the list of strategies slashing requested is being modified for
-     * @param bipsToIncrease the basis points slashing to modify for given strategies
+     * @dev pushes a MagnitudeUpdate and TotalAndNonslashableUpdate to take effect in 21 days
+     * @dev reverts if magnitudeDiff > allocated magnitude
+     * since you cannot deallocate more than is already allocated
+     * @dev emits event queueDeallocation
      */
-    function increaseRequestedBipsToSlash(
+    function queueDeallocation(
```
Comment (gpsanant) [inline src/contracts/core/Slasher.sol:128]: lowkey should we just do everything in ints so we dont duplicate this code twice

File: src/contracts/core/Slasher.sol [hunk 5]
```diff
@@ -5,233 +5,475 @@ import "./SlasherStorage.sol";
...
+    function _getLatestMagnitudeUpdate(
+        address operator,
+        IStrategy strategy,
+        OperatorSet calldata operatorSet
+    ) internal view returns (MagnitudeUpdate storage) {
+        uint256 magnitudeUpdatesLength = _magnitudeUpdates[operator][strategy][operatorSet.avs][operatorSet.id].length;
+        return _magnitudeUpdates[operator][strategy][operatorSet.avs][operatorSet.id][magnitudeUpdatesLength - 1];
+    }
+
+    /// TODO Handle reverting case if no updates exist, length == 0
```
Comment (gpsanant) [inline src/contracts/core/Slasher.sol:470]: lol 1 line todo for a 1 line change

File: src/contracts/core/Slasher.sol [hunk 6]
```diff
@@ -5,233 +5,475 @@ import "./SlasherStorage.sol";
...
+    function _checkPendingUpdates(
         address operator,
         IStrategy strategy,
-        uint32 epoch
-    ) external view returns (uint32) {
-        uint64 totalSlashingRate = slashingRequests[operator][strategy][epoch].slashingRate;
-        if (totalSlashingRate >= SlashingAccountingUtils.BIPS_FACTOR_SQUARED) {
-            totalSlashingRate = SlashingAccountingUtils.BIPS_FACTOR_SQUARED;
+        OperatorSet calldata operatorSet
+    ) public view {
```
Comment (gpsanant) [inline src/contracts/core/Slasher.sol:387]: public?

Comment (gpsanant) [reply src/contracts/core/Slasher.sol:387]: if we want to have this public, pls take of the `_`

Comment (gpsanant) [inline src/contracts/core/Slasher.sol:387]: stylistically, prefer if this returns a bool and we require

File: src/contracts/core/Slasher.sol [hunk 7]
```diff
@@ -5,233 +5,475 @@ import "./SlasherStorage.sol";
...
+        uint256 magnitudeUpdatesLength = _magnitudeUpdates[operator][strategy][operatorSet.avs][operatorSet.id].length;
+        if (magnitudeUpdatesLength == 0) {
+            return;
+        }
+        /// loop backwards from magnitude update history until either the the current magnitude is found
+        /// or 3 pending updates are found in wich case we revert.
+        for (uint256 i = magnitudeUpdatesLength - 1; i >= 0; --i) {
```
Comment (gpsanant) [inline src/contracts/core/Slasher.sol:394]: could we just do something like
```solidity
if (magnitudeUpdatesLength < MAX_PENDING_MAGNITUDE_UPDATES) {
    return true;
}

return _magnitudeUpdates[operator][strategy][operatorSet.avs][operatorSet.id][magnitudeUpdatesLength - MAX_PENDING_MAGNITUDE_UPDATES].timestamp < block.timestamp;
```

File: src/contracts/core/Slasher.sol [hunk 8]
```diff
@@ -275,94 +507,71 @@ contract Slasher is SlasherStorage {
...
+            slashingRate = slashingUpdates[operator][strategy][lookupEpoch].slashingRate;
+        }
+        return slashingRate;
+    }
+
     // @notice Returns the epoch in which the operator was last slashed
     function lastSlashed(address operator, IStrategy strategy) public view returns (uint32) {
-        uint256 slashedEpochHistoryLength = slashedEpochHistory[operator][strategy].length;
-        if (slashedEpochHistoryLength == 0) {
+        uint256 slashingEpochHistoryLength = slashingEpochHistory[operator][strategy].length;
```
Comment (gpsanant) [inline src/contracts/core/Slasher.sol:543]: no epochs

File: src/contracts/core/Slasher.sol [hunk 9]
```diff
@@ -275,94 +507,71 @@ contract Slasher is SlasherStorage {
...
+        IStrategy strategy,
+        uint32 epoch
+    ) internal view returns (uint32, bool) {
         // TODO DONE?: note the edge case of 0th epoch; need to make sure it's clear how it should be handled
         uint32 epochForLookup;
         bool found;
-        for (uint256 i = slashedEpochHistory[operator][strategy].length; i > 0; --i) {
-            epochForLookup = slashedEpochHistory[operator][strategy][i - 1];
+        for (uint256 i = slashingEpochHistory[operator][strategy].length; i > 0; --i) {
+            epochForLookup = slashingEpochHistory[operator][strategy][i - 1];
```
Comment (gpsanant) [inline src/contracts/core/Slasher.sol:569]: no epochs

File: src/contracts/core/Slasher.sol [hunk 10]
```diff
@@ -5,233 +5,475 @@ import "./SlasherStorage.sol";
...
-        uint32 epoch
-    ) public view returns (uint32) {
-        return requestedSlashedBips[operator][strategy][epoch][_hashOperatorSet(operatorSet)]
-            * operatorSetManager.getSlashableBips(operator, operatorSet, strategy, epoch);
+        OperatorSet calldata operatorSet,
+        MagnitudeUpdate storage magnitudeUpdate,
+        uint64 newMagnitude,
+        uint32 effectTimestamp
+    ) internal {
+        if (magnitudeUpdate.timestamp == effectTimestamp) {
```
Comment (gpsanant) [inline src/contracts/core/Slasher.sol:349]: i feel like making this check earlier and reverting if called twice in a block is better? idk

File: src/contracts/core/Slasher.sol [hunk 11]
```diff
@@ -5,233 +5,475 @@ import "./SlasherStorage.sol";
...
-     * operator, strategy, epoch, and operator set
-     *
-     * @dev may exceed the AVS operator set's allowed slashing per epoch;
-     * the `getPendingSlashedPPHM` will accurately reflect this ceiling though.
-     */
-    function getRequestedSlashingRate(
+    /// @notice push or update the latest magnitude update for the given operator, strategy, and operatorSet
+    /// depending on the current effectTimestamp
+    function _updateMagnitudeHistory(
         address operator,
```
Comment (gpsanant) [inline src/contracts/core/Slasher.sol:342]: you can pass storage mappings as parameters to internal functions. that would probably be cleaner here

File: src/contracts/core/Slasher.sol [hunk 12]
```diff
@@ -5,233 +5,475 @@ import "./SlasherStorage.sol";
...
-        uint32 requestedSlashingRate = getRequestedSlashingRate(operator, strategy, operatorSet, epoch);
-        if (requestedSlashingRate >= SlashingAccountingUtils.BIPS_FACTOR_SQUARED) {
-            requestedSlashingRate = uint32(SlashingAccountingUtils.BIPS_FACTOR_SQUARED);
+        TotalAndNonslashableUpdate storage totalAndNonslashableUpdate,
+        uint64 newTotalMagnitude,
+        uint64 newNonslashableMagnitude,
+        uint32 effectTimestamp
+    ) internal {
+        if (totalAndNonslashableUpdate.timestamp == effectTimestamp) {
+            totalAndNonslashableUpdate.totalMagnitude = newTotalMagnitude;
```
Comment (gpsanant) [inline src/contracts/core/Slasher.sol:370]: same comments as above

File: src/contracts/core/Slasher.sol (hunk 6 above)
Comment (8sunyuan) [reply src/contracts/core/Slasher.sol:387]: agreed

File: src/contracts/core/Slasher.sol (hunk 7 above)
Comment (8sunyuan) [reply src/contracts/core/Slasher.sol:394]: oh shit nice optimization

File: src/contracts/core/Slasher.sol (hunk 6 above)
Comment (8sunyuan) [reply src/contracts/core/Slasher.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/623/commits/f7c9a75affd75db460bf53ff9dc6d5b8ff9d32db

File: src/contracts/core/Slasher.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (8sunyuan) [reply src/contracts/core/Slasher.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/623/commits/f7c9a75affd75db460bf53ff9dc6d5b8ff9d32db
added

File: src/contracts/core/Slasher.sol (hunk 4 above)
Comment (8sunyuan) [reply src/contracts/core/Slasher.sol:124]: Tried to think of best way to reuse some kind of internal function.
Main thing with this is that addedMagnitude check is comparing with nonslashableMagnitude
and decrementedMagnitude check is comparing with latest allocated magnitude.
Handling these two different cases between allocation/deallocation seemed difficult so I left the params as uint64

File: src/contracts/core/Slasher.sol (hunk 3 above)
Comment (8sunyuan) [reply src/contracts/core/Slasher.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/623/commits/f7c9a75affd75db460bf53ff9dc6d5b8ff9d32db
Will add signature verification later

File: src/contracts/core/Slasher.sol (hunk 2 above)
Comment (8sunyuan) [reply src/contracts/core/Slasher.sol]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/623/commits/f7c9a75affd75db460bf53ff9dc6d5b8ff9d32db
Added allocator role in `OperatorDetails` struct

File: src/contracts/core/Slasher.sol (hunk 1 above)
Comment (8sunyuan) [reply src/contracts/core/Slasher.sol:28]: https://github.com/Layr-Labs/protocol-designs/pull/12/files#diff-ba96d21a88b00975a5f3e56038db353ef3764544cb431fcbae024e9726ba74f3R17-R20
Its up-to-date according to latest gitbook status

**Pull request closed without merging**